# Demand Forecasting (Sales Prediction)

## 1.Problem Statement

- In real-world businesses, one of the biggest challenges is:

“How many units will customers buy in the future?”

- Poor prediction leads to:

   - Overstock → wasted money
   - Understock → lost sales
   - Supply chain chaos

So the goal is to build a model that can accurately predict product sales (Quantity) using historical data.

## 2.Objectives
- Analyze sales dataset and uncover patterns
- Perform data cleaning & feature engineering
- Transform raw data into ML-ready format
- Train multiple regression models
- Compare models using R² & MAE
- Select best model with strong generalization
- Provide business insights from features

## 3.Import require libraries

In [11]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import AdaBoostRegressor

from xgboost import XGBRegressor

## 4.Load Dataset

In [3]:
df = pd.read_csv("Dataspy_dataset.csv")
df.head(8)

,Invoice_Date,Product,Region,Revenue,Quantity,Order_ID,Delivery_Date,Status,Stock
0,2025-04-01,Product_A,Nashik,12000,5,ORD001,2025-04-05,Delivered,40
1,2025-04-03,Product_B,Pune,18000,8,ORD002,2025-04-08,Delivered,55
2,2025-04-07,Product_C,Mumbai,9500,3,ORD003,2025-04-15,Delivered,30
3,2025-04-10,Product_D,Nagpur,22000,10,ORD004,2025-04-18,In Process,60
4,2025-04-15,Product_E,Aurangabad,15000,6,ORD005,2025-04-25,Delivered,45
5,2025-04-02,Product_F,Solapur,13000,5,ORD006,2025-04-12,Delayed,35
6,2025-04-20,Product_G,Kolhapur,20000,9,ORD007,2025-04-30,In Process,50
7,2025-04-25,Product_H,Amravati,11000,4,ORD008,2025-05-02,Delivered,25


## 5.Check Missing Values & Handle

In [4]:
df.isnull().sum()

Invoice_Date     0
Product          0
Region           0
Revenue          0
Quantity         0
Order_ID         0
Delivery_Date    0
Status           0
Stock            0
dtype: int64

## 6.Data Preprocessing

### 6.1 Convert Date Columns

In [5]:
df['Invoice_Date'] = pd.to_datetime(df['Invoice_Date'])
df['Delivery_Date'] = pd.to_datetime(df['Delivery_Date'])

### 6.2 Feature Engineering

In [6]:
df['invoice_day'] = df['Invoice_Date'].dt.day
df['invoice_month'] = df['Invoice_Date'].dt.month
df['delivery_days'] = (df['Delivery_Date'] - df['Invoice_Date']).dt.days

### 6.3 Drop Unnecessary Columns

In [7]:
df.drop(['Order_ID', 'Invoice_Date', 'Delivery_Date', 'Revenue'], axis=1, inplace=True)

### 6.4 Encode Categorical Data (ENCODING)

In [8]:
df = pd.get_dummies(df, drop_first=True)

## 7.Model Building

### 7.1 Define Features and Target

In [9]:
X = df.drop('Quantity', axis=1)
y = df['Quantity']

### 7.2 Train-Test Split

In [10]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# 1.Linear Regression

## Before Tunning

In [20]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error

# Create model
lr = LinearRegression()

# Train
lr.fit(X_train, y_train)

# Predict
y_pred = lr.predict(X_test)

# Evaluate
print("Before Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))

Before Tuning
R2: 1.0
MAE: 6.0840221749458575e-15


## After Tunning

In [21]:
from sklearn.model_selection import GridSearchCV

params = {
    "fit_intercept": [True, False]
}

grid = GridSearchCV(
    LinearRegression(),
    params,
    cv=5,
    scoring='r2'
)

# Train
grid.fit(X_train, y_train)

# Best model
best_lr = grid.best_estimator_

# Predict
y_pred = best_lr.predict(X_test)

# Evaluate
print("\nAfter Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))


After Tuning
R2: 1.0
MAE: 6.0840221749458575e-15


# 2.Decision Tree

## Before Tunning

In [22]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score, mean_absolute_error

# Create model
dt = DecisionTreeRegressor(random_state=42)

# Train
dt.fit(X_train, y_train)
# Predict
y_pred = dt.predict(X_test)

# Evaluate
print("Before Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))


Before Tuning
R2: 0.9189189189189189
MAE: 0.15


## After Tunning

In [23]:
from sklearn.model_selection import GridSearchCV

# Parameters to test
params = {
    "max_depth": [3, 5, 10],
    "min_samples_split": [2, 5, 10]
}

# GridSearch
grid = GridSearchCV(
    DecisionTreeRegressor(random_state=42),
    params,
    cv=5,
    scoring='r2'
)

# Train
grid.fit(X_train, y_train)

# Best model
best_dt = grid.best_estimator_

# Predict
y_pred = best_dt.predict(X_test)

# Evaluate
print("\nAfter Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))



After Tuning
R2: 0.9189189189189189
MAE: 0.15


# 3.Random Forest

## Before Tunning

In [24]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score, mean_absolute_error

# Create model
rf = RandomForestRegressor(random_state=42)

# Train
rf.fit(X_train, y_train)

# Predict
y_pred = rf.predict(X_test)

# Evaluate
print("Before Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))

Before Tuning
R2: 0.9557018018018018
MAE: 0.2185


## After Tunning

In [25]:
from sklearn.model_selection import GridSearchCV

# Parameters to test
params = {
    "n_estimators": [50, 100],
    "max_depth": [5, 10, None]
}

# GridSearch
grid = GridSearchCV(
    RandomForestRegressor(random_state=42),
    params,
    cv=5,
    scoring='r2'
)

# Train
grid.fit(X_train, y_train)

# Best model
best_rf = grid.best_estimator_

# Predict
y_pred = best_rf.predict(X_test)

# Evaluate
print("\nAfter Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))


After Tuning
R2: 0.9434810810810811
MAE: 0.244


# 4.AdaBoost 

## Before Tunning

In [26]:
from sklearn.ensemble import AdaBoostRegressor
from sklearn.metrics import r2_score, mean_absolute_error

# Create model
ada = AdaBoostRegressor(random_state=42)

# Train
ada.fit(X_train, y_train)

# Predict
y_pred = ada.predict(X_test)

# Evaluate
print("Before Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))

Before Tuning
R2: 0.9712651994398026
MAE: 0.17023809523809522


## After Tunning

In [27]:
from sklearn.model_selection import GridSearchCV

# Parameters to test
params = {
    "n_estimators": [50, 100, 200],
    "learning_rate": [0.01, 0.1, 1]
}

# GridSearch
grid = GridSearchCV(
    AdaBoostRegressor(random_state=42),
    params,
    cv=5,
    scoring='r2'
)

# Train
grid.fit(X_train, y_train)

# Best model
best_ada = grid.best_estimator_

# Predict
y_pred = best_ada.predict(X_test)

# Evaluate
print("\nAfter Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))


After Tuning
R2: 0.9685593298503139
MAE: 0.20133175158175157


# 5.XGBoost

## Before Tunning

In [28]:
from xgboost import XGBRegressor
from sklearn.metrics import r2_score, mean_absolute_error

# Create model
xgb = XGBRegressor(random_state=42)

# Train
xgb.fit(X_train, y_train)

# Predict
y_pred = xgb.predict(X_test)

# Evaluate
print("Before Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))

Before Tuning
R2: 0.9420295357704163
MAE: 0.12891967594623566


## After Tunning

In [29]:
from sklearn.model_selection import GridSearchCV

# Parameters to test
params = {
    "n_estimators": [50, 100],
    "learning_rate": [0.05, 0.1],
    "max_depth": [3, 5],
    "subsample": [0.8, 1]
}

# GridSearch
grid = GridSearchCV(
    XGBRegressor(random_state=42),
    params,
    cv=5,
    scoring='r2'
)

# Train
grid.fit(X_train, y_train)

# Best model
best_xgb = grid.best_estimator_

# Predict
y_pred = best_xgb.predict(X_test)

# Evaluate
print("\nAfter Tuning")
print("R2:", r2_score(y_test, y_pred))
print("MAE:", mean_absolute_error(y_test, y_pred))


After Tuning
R2: 0.9817751049995422
MAE: 0.1199168786406517


## 8.Final Conclusion

The project focused on predicting sales (Quantity) using multiple regression models. Different models were trained and evaluated using R² score and MAE to ensure both accuracy and real-world error understanding.

Among all models, XGBoost (after tuning) delivered the best performance with the highest R² (0.98) and lowest MAE (0.11), making it the most reliable and accurate model for this dataset.

Linear Regression showed a perfect R² score but a high MAE, indicating misleading performance and possible overfitting or data leakage, hence it was not considered reliable.

Decision Tree, Random Forest, and AdaBoost performed well, but either had slightly higher error or did not improve after tuning. This highlights that ensemble methods are powerful, but proper tuning is critical for optimal performance.

Overall, the project demonstrates that:

Proper feature engineering and model comparison are essential
Evaluating with both R² and MAE prevents wrong conclusions
XGBoost is highly effective for structured regression problems